From f7a6849fb0312c0d87b77b771f915a192ab5b4f3 Mon Sep 17 00:00:00 2001
From: Naga Sureshkumar Relli <naga.sureshkumar.relli@xilinx.com>
Date: Thu, 2 Feb 2017 10:41:06 +0530
Subject: [PATCH 1221/1566] mtd: nand: arasan_nand: Change ref clk in SDR
 modes 2 to 5 to less than 90MHz

commit  cab82219feca8fc6a1758c521f64afea49a14918 from
https://github.com/Xilinx/linux-xlnx.git

Since arasan nand SDR modes 2 to 5 are failing to capture the read
data at 90MHz. so change the nand_ref_clk to < 90MHz, if SDR timing
modes are mode 2 to mode 5 and 80MHz is chosen as reasonable number.

Signed-off-by: Naga Sureshkumar Relli <nagasure@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/nand/arasan_nand.c |   23 +++++++++++++++++++++++
 1 files changed, 23 insertions(+), 0 deletions(-)

diff --git a/drivers/mtd/nand/arasan_nand.c b/drivers/mtd/nand/arasan_nand.c
index 4ab4cd4..80fa7a2 100644
--- a/drivers/mtd/nand/arasan_nand.c
+++ b/drivers/mtd/nand/arasan_nand.c
@@ -99,6 +99,8 @@
 #define ONFI_DATA_INTERFACE_NVDDR      BIT(4)
 #define EVENT_MASK	(XFER_COMPLETE | READ_READY | WRITE_READY | MBIT_ERROR)
 
+#define SDR_MODE_DEFLT_FREQ		80000000
+
 /**
  * struct anfc_nand_chip - Defines the nand chip related information
  * @node:		used to store NAND chips into a list.
@@ -719,6 +721,7 @@ static int anfc_init_timing_mode(struct anfc *nfc,
 	u32 inftimeval;
 	struct nand_chip *chip = &achip->chip;
 	struct mtd_info *mtd = nand_to_mtd(chip);
+	bool change_sdr_clk = false;
 
 	memset(feature, 0, NVDDR_MODE_PACKET_SIZE);
 	/* Get nvddr timing modes */
@@ -726,6 +729,8 @@ static int anfc_init_timing_mode(struct anfc *nfc,
 	if (!mode) {
 		mode = fls(onfi_get_async_timing_mode(chip)) - 1;
 		inftimeval = mode;
+		if (mode >= 2 && mode <= 5)
+			change_sdr_clk = true;
 	} else {
 		mode = fls(mode) - 1;
 		inftimeval = NVDDR_MODE | (mode << NVDDR_TIMING_MODE_SHIFT);
@@ -740,6 +745,24 @@ static int anfc_init_timing_mode(struct anfc *nfc,
 	if (err)
 		return err;
 
+	/*
+	 * SDR timing modes 2-5 will not work for the arasan nand when
+	 * freq > 90 MHz, so reduce the freq in SDR modes 2-5 to < 90Mhz
+	 */
+	if (change_sdr_clk) {
+		clk_disable_unprepare(nfc->clk_sys);
+		err = clk_set_rate(nfc->clk_sys, SDR_MODE_DEFLT_FREQ);
+		if (err) {
+			dev_err(nfc->dev, "Can't set the clock rate\n");
+			return err;
+		}
+		err = clk_prepare_enable(nfc->clk_sys);
+		if (err) {
+			dev_err(nfc->dev, "Unable to enable sys clock.\n");
+			clk_disable_unprepare(nfc->clk_sys);
+			return err;
+		}
+	}
 	achip->inftimeval = inftimeval;
 
 	if (mode & ONFI_DATA_INTERFACE_NVDDR)
-- 
1.7.5.4

